src/Approximations/overapproximate.jl

```diff
@@ -516,3 +516,93 @@ function overapproximate(cap::Intersection{N,
     return overapproximate(swap(cap), dir; kwargs...)
 end
 
+"""
+    overapproximate(lm::LinearMap{N, <:CartesianProductArray{N, <:LazySet{N}}},
+                         dir::Type{<:AbstractDirections})::CartesianProductArray{N, HPolytope{N}} where {N}
+
+Overapproximating a lazy linear map of cartesian product array with template directions for each block.
+
+### Input
+
+- `lm`   -- lazy linear map of cartesian product array
+- Type{<:CartesianProductArray} -- type to specify decomposed overapproximation
+- `oa` -- (concrete) direction representation
+
+### Output
+
+An `CartesianProductArray` with overapproximation of the each block with the directions from `dir`.
+"""
+function overapproximate(lm::LinearMap{N, <:CartesianProductArray{N,
+                            <:LazySet{N}}}, ::Type{<:CartesianProductArray}, oa::Union{Type{<:AbstractDirections},Type{<:Hyperrectangle}, Type{<:Interval}}=Hyperrectangle) where {N}
+
+    if !(size(lm.M, 2) == dim(lm.X))
+        error("Matrix needs to be commensurate with the cartesian product")
+    end
+
+    cp = lm.X
+    M = lm.M
+
+    col_st_index, col_end_ind = 0, 0
+
+    return decomposed_overapproximation(cp, M, col_st_index, col_end_ind, oa)
+end
+
+#template directions
+function decomposed_overapproximation(cp::CartesianProductArray{N,<:LazySet{N}}, M::AbstractArray{N},
+            col_st_index::Int, col_end_ind::Int, oa::Type{<:AbstractDirections}) where {N}
```

Comment:
There is no need for different methods, so you can move that code back to the main function. Just replace `oa{N}(dim(v_block))` by `oa` and `CartesianProductArray{N, HPolytope{N}}(array)` by `CartesianProductArray(array)` below.

One downside is that you cannot preallocate the `array`. So maybe you want to outsource this particular line:
```julia
function overapproximate(...)
    ...
    array = allocate_result(oa, N)
    ...
end

function allocate_result(oa, N)
    return Vector{LazySet{N}}()
end

function allocate_result(oa::Type{<:LazySet}, N)
    return Vector{oa{N}}()
end

function allocate_result(oa::oa::Type{<:AbstractDirections}, N)
    return Vector{HPolytope{N}}()
end
```

Reply:
Why do we need:

```
function allocate_result(oa, N)
    return Vector{LazySet{N}}()
end
```
?

Reply:
This is a fallback for other overapproximations.